core/src/wallet/cosmos/explorers/GaiaCosmosLikeBlockchainExplorer.cpp

```diff
@@ -427,5 +433,356 @@ namespace ledger {
+
+                auto fees = makeAmountArray(fees_.amount, allocator);
+                baseReq.AddMember(cosmos::constants::kFees, fees, allocator);
+
+                auto simulate = rapidjson::Value(rapidjson::kTrueType);
+                baseReq.AddMember(cosmos::constants::kSimulate, simulate, allocator);
+
+                return baseReq;
+            }
+
+        private:
+            std::string from_;
+            std::string memo_;
+            std::string chainId_;
+            std::string accountNumber_;
+            std::string accountSequence_;
+            std::string gas_;
+            std::string gasAdjustment_;
+            cosmos::Fee fees_;
+            bool simulate_;
+        };
+
+        rapidjson::Value makeBaseReq(const std::shared_ptr<api::CosmosLikeTransaction>& transaction,
+                                     const std::shared_ptr<api::CosmosLikeMessage>& message,
+                                     rapidjson::Document::AllocatorType& allocator)
+        {
+            const auto tx = std::dynamic_pointer_cast<CosmosLikeTransactionApi>(transaction);
+            const auto msg = std::dynamic_pointer_cast<CosmosLikeMessage>(message);
+            const auto baseReq = BaseReq(
+                    msg->getFromAddress(),
+                    tx->getMemo(),
+                    networks::getCosmosLikeNetworkParameters(tx->getCurrency().name).ChainId,
+                    tx->getAccountNumber(),
+                    tx->getAccountSequence(),
+                    tx->getTxData().fee.gas.toString(),
+                    std::string("1"),
+                    tx->getTxData().fee,
+                    true);
+
+            return baseReq.toJson(allocator);
+        }
+
+        struct JsonObject
+        {
+            rapidjson::GenericStringRef<char> name;
+            rapidjson::Value& value;
+
+            JsonObject(const char* name, rapidjson::Value& value)
+                : name(rapidjson::StringRef(name)), value(value)
+            {}
+        };
+
+        template <typename... JsonObj>
+        std::string makeJsonFrom(rapidjson::Document& document,
+                                 JsonObj&&... jsonObject)
+        {
+            auto& allocator = document.GetAllocator();
+
+            // TODO: use fold expression when C++17
+            // (document.AddMember(std::forward<JsonObj>(jsonObject).name, std::forward<JsonObj>(jsonObject).value, allocator), ...);
+            using _t = int[];
+            (void)_t{0, (document.AddMember(std::forward<JsonObj>(jsonObject).name,
+                        std::forward<JsonObj>(jsonObject).value, allocator), 0)...};
+
+            auto buffer = rapidjson::StringBuffer();
+            auto writer = rapidjson::Writer<rapidjson::StringBuffer>(buffer);
+            // TODO: sortJson is defined static in CosmosLikeTransactionApi.cpp (either make it visible or move this to CosmosLikeTransactionApi)
+            // sortJson(document);
```

Comment:
Since the device does not need to sign the payload here, sorting the json should not matter here.

But having sortJson as a hidden function is not so good I agree. Maybe we'll move it up to `utils` or something.